### ISSUE TITLE
Introduce a new datastore for UI state

### ISSUE THREAD

**Issue by eugene-manuilov** [google/site-kit-wp#2456]
## Feature Description

After discussing #2392 on our internal call, we agreed that it would be more convenient to create a completely new datastore to let different logical components to store UI related state there. This new datastore should become our go-to place when it comes to knowing the selected date range, which tab is open in a widget, or which metric is select for a graph.

Having UI information stored in the new datastore eliminates the weirdness that we have right now when that information is stored across different stores that are not intended for such information. Plus, it gives us a good basis for the next UI/UX enhancements that we can implement having this datastore. For example, we can save information stored in this datastore in persistent storage (for example, in the IndexedDB) and load it on the next visit to restore UI from the last session.

The datastore itself should be pretty simple at the beginning: it should be key-value storage with appropriate setValue action and getValue selector. All additional features can be implemented later on.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a new `core/ui` JS datastore with selector `getValue( key )` and action `setValue( key, value )`. It should also have a `setValues( values )` that allows to set multiple key-value pairs at a time.
    * Essentially, it should be a simple key-value storage, almost similar to the `core/forms` store but without the extra nesting of having key-value pairs as part of a specific form.
* The new Analytics All Traffic widget and the PageSpeed Insights widget should have their implementation modified to rely on the `core/ui` store instead of the `core/forms` store, since their interactions are not really based on form content. The keys used should be:
    * `dashboardAllTrafficWidgetDimensionName`
    * `dashboardAllTrafficWidgetDimensionValue`
    * `dashboardPageSpeedWidgetStrategy`
    * `dashboardPageSpeedWidgetDataSource`

## Implementation Brief

* Create `assets/js/googlesitekit/datastore/ui/constants.js` file:
  * Export a new constant `STORE_NAME` with `core/ui` value;
  * Export the same `STORE_NAME` constant as `CORE_UI` alias.
* Create `assets/js/googlesitekit/datastore/ui/ui.js` file:
  * Define the standard structure of the regular datastore using [forms](https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/datastore/forms/forms.js) datastore as an example;
  * Add `setValue( key, value )` and `setValues( values )` actions with appropriate reducers that add provided data into the state object;
  * Add `getValue( key )` selector that returns data from the state using provided key.
* Create `assets/js/googlesitekit/datastore/ui/index.js` file:
  * It should follow the similar forms [index.js](https://github.com/google/site-kit-wp/blob/develop/assets/js/googlesitekit/datastore/forms/index.js) file but use the ui datastore, define above, along with the `Data.commonStore` datastore;
  * Register the combined store using the `Data.registerStore` function and `STORE_NAME` constant imported from the `constants.js` file.
* Create `assets/js/googlesitekit-datastore-ui.js` file:
  * It should import the entire `assets/js/googlesitekit/datastore/ui/index.js` file.
* Edit the webpack config file to add a new entry point `googlesitekit-datastore-ui` that points to the `assets/js/googlesitekit-datastore-ui.js` file;
* Add the new Script for the new core/ui datastore to the assets list in the `includes/Core/Assets/Assets.php` file (see the `get_assets` method):
  * Use `googlesitekit-datastore-ui` handler for the new script and add this handler to the `$dependencies` array;
  * Add `googlesitekit-data` as a dependency to the new script.
* Update the new All Traffic widget to use the `core/ui` datastore instead of `core/forms` to control the current values for `dimensionName`, `dimensionValue`, and `dimensionColor` variables. Use the following keys for those variables:
  * `dashboardAllTrafficWidgetDimensionName`
  * `dashboardAllTrafficWidgetDimensionValue`
  * `dashboardAllTrafficWidgetDimensionColor`
* Update the PageSpeed Insights widget to use the `core/ui` datastore instead of `core/forms` to control the current values for `strategy` and `dataSrc` variables. Use the following keys for those variables:
  * `dashboardPageSpeedWidgetStrategy`
  * `dashboardPageSpeedWidgetDataSource`

### Test Coverage

* Add javascript tests to check the new actions and selectors of the `core/ui` datastore.

### Visual Regression Changes

* No visual changes are expected as a result of this change;

## QA Brief

* Make sure that there is a new `core/ui` datastore and it is used by the new All Traffic widget and the PSI widget to store user-selected values.

## Changelog entry

* Introduce `core/ui` datastore for centrally storing UI state in JS.

**Comment by felixarntz** [comment]
@eugene-manuilov IB pretty much ✅ , however I don't think you need to put anything under "Visual Regression Changes", they shouldn't be affected. What you have there furthermore is not a requirement for this issue and shouldn't be the case. For now, the `core/ui` store shouldn't have any persistence - we can consider that for the future, but for now it should be just another place than `core/forms` to store key-value pairs, for places which are not forms.

In other words, this issue should not result in _any_ user-facing changes.

**Comment by eugene-manuilov** [comment]
@felixarntz, gotcha! Updated.

**Comment by felixarntz** [comment]
IB ✅

**Comment by johnPhillips** [comment]
✅  QA
